text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,233 @@
+
+## Launching the Gateway
+
+#### Standalone process
+When running standalone, the Gateway is effectively a 'client' application to a set of peers.  A gateway instance is associated with an organization and will be configured to connect to a peer in the same organization in order to invoke discovery.  To do this it will need a signing identity in the channel writers policy.
+
+To run the gateway server, the following parameters must to be supplied:
+- The url of at least one peer in the org.  Once connected, the discovery service will be invoked to find other peers.
+- The MSPID associated with the gateway.
+- The signing identity of the gateway (cert and key), e.g. location of PEM files on disk.  
+- The CA certificate so that the gateway can make TLS connections to other components (peers/orderers) in the organization
+
+Note that the signing identity is required so the gateway server can make discovery requests and to register event listeners.  It is not used for submitting any transactions on behalf of end users.
+
+#### Embedded in Peer
+When embedded in a peer, the gateway will register its gRPC service with the host peer's gRPC server.
+The gateway does not need its own (client) identity since it can get all the discovery and event information directly from the host peer rather than via a gRPC signed request.
+
+### Discovery
+
+On startup, the gateway will connect to its primary peer(s), as specified in the command line option (or its host, if embedded in a peer).  The discovery service will be invoked to build and maintain a cache of the network topology per channel.  This will be used to identify endorsers based on the chaincode endorsement policy.
+
+#### Collection endorsement policy
+
+Discovery can be used to get endorsing peers for a given chaincode where the transaction function will write to a given collection, and this will take into account the collection level endorsement policy.
+Ideally we could take care of these endorsement requirements in the Gateway without the client needing to have any implicit knowledge of them. Suggest extending contract metadata to include information on collections, which can be queried by Fabric Gateway. If that isn't possible then we might need some way for the client to specify the collections that will be used (not desirable).
+
+### State based endorsement
+
+This is driven by business logic and rules. Requires the client to supply constraints on endorsing organisations as part of the proposal endorsement.
+
+### Chaincode-to-chaincode endorsement
+
+Augment contract metadata for transactions with additional chaincodes / transactions that will be invoked, then we use discovery to narrow the set of endorsers.
```

Comment:
I wonder how this is implemented today. Thanks in advance if a refence link to docs/git commit can be shared here.